presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitSource.java

```diff
@@ -262,6 +273,89 @@ public boolean isFinished(OptionalInt bucketNumber)
                 highMemorySplitSourceCounter);
     }
 
+    public static HiveSplitSource bucketedRewindable(
+            ConnectorSession session,
+            String databaseName,
+            String tableName,
+            TupleDomain<? extends ColumnHandle> compactEffectivePredicate,
+            int maxInitialSplits,
+            DataSize maxOutstandingSplitsSize,
+            HiveSplitLoader splitLoader,
+            Executor executor,
+            CounterStat highMemorySplitSourceCounter)
+    {
+        AtomicReference<State> stateReference = new AtomicReference<>(State.initial());
+        return new HiveSplitSource(
+                session,
+                databaseName,
+                tableName,
+                compactEffectivePredicate,
+                new PerBucket()
+                {
+                    private final Map<Integer, List<InternalHiveSplit>> splits = new ConcurrentHashMap<>();
```

Comment:
I just realized that all the methods in this inner class have to be synchronized, as the inner list is a simple non concurrent ArrayList.

Let's replace this map with a regular `HashMap` to avoid confusion, and mark it as `@GuardedBy("this")`, and synchronize `offer`, `borrowBatchAsync` and `rewind`

Reply:
In theory we don't need to synchronize `offer` because `rewind` and `borrowBatchAsync` will be blocked before all `offer` calls finish, but yeah let's make them all synchronized to be safe.

Reply:
Yeah, you are right. But yes, just to be safe.